Libraries/MLXLLM/Models/Gemma2.swift

```diff
@@ -288,3 +289,120 @@ extension Gemma2Model: LoRAModel {
     }
 }
+
+// TODO - notes
+//
+// - make UnaryLayer extend Module
+// - make a Quantized protocol that provides the groupSize and bits
+// - make the QuantizedLinear shape produce the expanded shape
+// - make `items()` open
+// - make `updateModule(key:_:)` open
+//
+// - evaluation and training should work as expected
+// - this flattens the weights and modules into one layer
+//      - to match the normal lora implementation
+//      - see items() and updateModule()
+
+// TODO: make UnaryLayer extend Module
+public protocol UnaryLayer2: Module {
+    func callAsFunction(_ x: MLXArray) -> MLXArray
+}
+
+/// LoRA layer that can wrap any UnaryLayer
+class LoRA: Module, UnaryLayer2 {
+
+    let adapts: UnaryLayer2
+    let scale: Float
+
+    @ParameterInfo(key: "lora_a") var loraA: MLXArray
+    @ParameterInfo(key: "lora_b") var loraB: MLXArray
+
+    public init(
+        adapts: UnaryLayer2, inputDimensions: Int, outputDimensions: Int, rank: Int = 8,
+        scale: Float = 20.0
+    ) {
+        self.adapts = adapts
+
+        self.scale = scale
+
+        let loraScale = 1 / sqrt(Float(inputDimensions))
+        self._loraA.wrappedValue = MLXRandom.uniform(
+            low: -loraScale, high: loraScale, [inputDimensions, rank])
+        self._loraB.wrappedValue = MLXArray.zeros([rank, outputDimensions])
+
+        freeze()
+    }
+
+    // TODO: in LoRALinear this is
+    // public static func from(linear: Linear, rank: Int = 8) -> LoRA
+    public convenience init(linear: Linear, rank: Int = 8, scale: Float = 20.0) {
+        var (outputDimensions, inputDimensions) = linear.shape
+
+        if let linear = linear as? QuantizedLinear {
+            // TODO Linear should probably have a property to return these directly
+            // rather than shape which represents the physical shape of the layers
+            inputDimensions = inputDimensions * 32 / linear.bits
+        }
+
+        self.init(
+            adapts: linear,
+            inputDimensions: inputDimensions, outputDimensions: outputDimensions,
+            rank: rank, scale: scale)
+    }
+
+    // produce a merged view of properties (flatten LoRA into adapts)
+    override func items() -> ModuleItems {
+        var result = adapts.items()
+        for (key, value) in super.items() {
+            if key == "adapts" { continue }
+            result[key] = value
+        }
+        return result
+    }
+
+    // forward module updates -> adapt
+    func updateModule(key: String, _ value: Any) throws {
+        try adapts.updateModule(key: key, value)
+    }
+
+    override func freeze(recursive: Bool = true, keys: [String]? = nil, strict: Bool = false) throws
+    {
+        try adapts.freeze(recursive: recursive, keys: keys, strict: strict)
+    }
+
+    // TODO: this requires knowledge of the innards of the adapted layer so it
+    // is specific to Linear (and QuantizedLinear).
+    public func toLinear(deQuantize: Bool = false) -> Linear {
```

Comment:
The `fuse` operation requires knowledge of how to combine the LoRA weights with the target.  Type-wise this could easily return a `UnaryLayer` but it has to understand the implementation in order to fuse.